<?xml version="1.0"?>
<robot name="robor" xmlns:xacro="http://ros.org/wiki/xacro">

  <!-- Common Properties -->
    <xacro:property name="wheel_radius" value="0.1"/>
    <xacro:property name="wheel_length" value="0.05"/>
    <xacro:property name="body_length" value="1.0"/>
    <xacro:property name="body_width" value="0.5"/>
    <xacro:property name="body_height" value="0.2"/>
    <xacro:property name="sensor_box_size" value="0.1"/>
    <xacro:property name="imu_box_size" value="0.05"/>

    <link name="world"/>

    <joint name="world_to_base_link" type="">
        <parent link="world"/>
        <child link="base_link"/>
        <origin xyz="0 0 0.1" rpy="0 0 0"/>
    </joint>


    <!-- Main Body -->
    <link name="base_link">
    <inertial>
        <origin xyz="0 0 0.1"/>
        <mass value="5.0"/>
        <inertia ixx="1.0" ixy="0.0" ixz="0.0" iyy="1.0" iyz="0.0" izz="1.0"/>
    </inertial>

        <collision name="base_link_collision">
            <geometry>
            <box size="${body_length} ${body_width} ${body_height}"/>
            </geometry>
            <origin xyz="0 0 0.5"/>
        </collision>
        <visual>
            <geometry>
            <box size="${body_length} ${body_width} ${body_height}"/>
            </geometry>
            <material name="blue">
            <color rgba="0 0 0.75 1"/>
            </material>
            <origin xyz="0 0 0.05" rpy="0 0 0"/>
        </visual>
    </link>


            <!-- Wheels -->
    <link name="wheel1">
        <inertial>
        <origin xyz="0 0 0"/>
        <mass value="0.5"/>
        <inertia ixx="0.005" ixy="0.0" ixz="0.0" iyy="0.005" iyz="0.0" izz="0.005"/>
        </inertial>
        <collision name="wheel1_collision">
        <geometry>
            <cylinder radius="${wheel_radius}" length="${wheel_length}"/>
        </geometry>
        <origin xyz="0 0 0" rpy="1.57 0 0"/>
        <surface>
            <friction>
            <ode>
                <mu>1.0</mu>
                <mu2>1.0</mu2>
            </ode>
            </friction>
        </surface>
        </collision>
        <visual>
        <geometry>
            <cylinder radius="${wheel_radius}" length="${wheel_length}"/>
        </geometry>
        <material name="red">
            <color rgba="0.75 0 0 1"/>
        </material>
        <origin xyz="0 0 0" rpy="1.57 0 0"/>
        </visual>
    </link>

    <!-- Wheel 2 -->
    <link name="wheel2">
        <inertial>
        <origin xyz="0 0 0"/>
        <mass value="0.5"/>
        <inertia ixx="0.005" ixy="0.0" ixz="0.0" iyy="0.005" iyz="0.0" izz="0.005"/>
        </inertial>
        <collision name="wheel2_collision">
        <geometry>
            <cylinder radius="${wheel_radius}" length="${wheel_length}"/>
        </geometry>
        <origin xyz="0 0 0" rpy="1.57 0 0"/>
        <surface>
            <friction>
            <ode>
                <mu>1.0</mu>
                <mu2>1.0</mu2>
            </ode>
            </friction>
        </surface>
        </collision>
        <visual>
        <geometry>
            <cylinder radius="${wheel_radius}" length="${wheel_length}"/>
        </geometry>
        <material name="red">
            <color rgba="0.75 0 0 1"/>
        </material>
        <origin xyz="0 0 0" rpy="1.57 0 0"/>
        </visual>
    </link>



    <!-- Camera -->
    <link name="camera_link">
        <visual>
        <geometry>
            <box size="${sensor_box_size} ${sensor_box_size} ${sensor_box_size}"/>
        </geometry>
        <material name="green">
            <color rgba="0 1 0 1"/>
        </material>
        <origin xyz="0.5 0 0.3" rpy="0 0 0"/>
        </visual>
    </link>

    <!-- Lidar -->
    <link name="lidar_link">
        <visual>
        <geometry>
            <cylinder radius="${wheel_radius}" length="${wheel_length}"/>
        </geometry>
        <material name="yellow">
            <color rgba="1 1 0 1"/>
        </material>
        <origin xyz="0.5 0 0.15" rpy="0 0 0"/>
        </visual>
    </link>

    <!-- IMU -->
    <link name="imu_link">
        <visual>
        <geometry>
            <box size="${imu_box_size} ${imu_box_size} ${imu_box_size}"/>
        </geometry>
        <material name="orange">
            <color rgba="1 0.5 0 1"/>
        </material>
        <origin xyz="0 0 0.3" rpy="0 0 0"/>
        </visual>
    </link>

  <!-- Joints -->
    <joint name="base_link_wheel1" type="continuous">
        <parent link="base_link"/>
        <child link="wheel1"/>
        <origin xyz="0 -0.25 0" rpy="0 0 0"/>
        <axis xyz="0 1 0"/>
        <limit effort="10.0" velocity="10.0"/>
        <dynamics damping="0.1" friction="0.1"/>
        </joint>

        <joint name="base_link_wheel2" type="continuous">
        <parent link="base_link"/>
        <child link="wheel2"/>
        <origin xyz="0 0.25 0" rpy="0 0 0"/>
        <axis xyz="0 1 0"/>
        <limit effort="10.0" velocity="10.0"/>
        <dynamics damping="0.1" friction="0.1"/>
    </joint>


    <joint name="base_link_camera" type="fixed">
        <parent link="base_link"/>
        <child link="camera_link"/>
    </joint>

    <joint name="base_link_lidar" type="fixed">
        <parent link="base_link"/>
        <child link="lidar_link"/>
    </joint>

    <joint name="base_link_imu" type="fixed">
        <parent link="base_link"/>
        <child link="imu_link"/>
    </joint>

    <transmission name="base_link_wheel1_trans">
        <type>transmission_interface/SimpleTransmission</type>
        <joint name="base_link_wheel1">
            <hardwareInterface>EffortJointInterface</hardwareInterface>
        </joint>
        <actuator name="wheel1_motor">
            <mechanicalReduction>1.0</mechanicalReduction>
        </actuator>
        </transmission>

        <transmission name="base_link_wheel2_trans">
        <type>transmission_interface/SimpleTransmission</type>
        <joint name="base_link_wheel2">
            <hardwareInterface>EffortJointInterface</hardwareInterface>
        </joint>
        <actuator name="wheel2_motor">
            <mechanicalReduction>1.0</mechanicalReduction>
        </actuator>
    </transmission>

  <!-- Plugins -->
    <gazebo>
        <plugin name="diff_drive_controller" filename="libgazebo_ros_diff_drive.so">
            <left_joint>base_link_wheel1</left_joint>
            <right_joint>base_link_wheel2</right_joint>
            <wheel_separation>0.5</wheel_separation>
            <wheel_radius>0.1</wheel_radius>
            <wheel_diameter>0.2</wheel_diameter>
            <torque>1.0</torque>
            <update_rate>50</update_rate>
            <cmd_vel_topic>cmd_vel</cmd_vel_topic>
            <odom_topic>odom</odom_topic>
            <odometry_frame>odom</odometry_frame>
            <robot_base_frame>base_link</robot_base_frame>
            <publish_wheel_tf>false</publish_wheel_tf>
            <publish_odom_tf>true</publish_odom_tf>
            <update_rate>30</update_rate>
        </plugin>
    </gazebo>


    <gazebo>
        <plugin name="camera_plugin" filename="libgazebo_ros_camera.so">
        <robotNamespace>/</robotNamespace>
        <cameraName>camera</cameraName>
        <frameName>camera_link</frameName>
        <imageTopicName>camera/image_raw</imageTopicName>
        <cameraInfoTopicName>camera/camera_info</cameraInfoTopicName>
        <updateRate>30</updateRate>
        </plugin>
    </gazebo>

    <gazebo>
        <plugin name="lidar_plugin" filename="libgazebo_ros_laser.so">
        <robotNamespace>/</robotNamespace>
        <frameName>lidar_link</frameName>
        <topicName>scan</topicName>
        <updateRate>30</updateRate>
        <rangeMin>0.12</rangeMin>
        <rangeMax>3.5</rangeMax>
        <samples>360</samples>
        <resolution>1.0</resolution>
        </plugin>
    </gazebo>

    <gazebo>
        <plugin name="imu_plugin" filename="libgazebo_ros_imu_sensor.so">
        <robotNamespace>/</robotNamespace>
        <bodyName>imu_link</bodyName>
        <updateRate>200</updateRate>
        <topicName>imu/data</topicName>
        <frameName>imu_link</frameName>
        </plugin>
    </gazebo>

</robot>
